File: algorithms/[739]每日温度/solution.py
# topics = ["栈"]

from typing import List


class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        n = len(T)
        res: List[int] = [0] * n

        # 单调栈
        stack: List[int] = []

        for j in range(n):
            while stack and T[stack[-1]] < T[j]:
                i = stack.pop()
                res[i] = j - i
            stack.append(j)

        return res

File: algorithms/[739]每日温度/Solution.java
// topics = ["栈"]

import java.util.Deque;
import java.util.LinkedList;

class Solution {
    /**
     * 单调栈
     * time O(n), space O(n), n 为数组长度
     */
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<Integer> stk = new LinkedList<>();
        int n = temperatures.length;
        int[] res = new int[n];

        for (int i = 0; i < n; i++) {
            while (stk.size() > 0 && temperatures[stk.peek()] < temperatures[i]) {
                int idx = stk.pop();
                res[idx] = i - idx;
            }
            stk.push(i);
        }

        return res;
    }
}